From 86f662a8202408134a235572ec60141d3082f975 Mon Sep 17 00:00:00 2001
From: "Jose E. Marchesi" <jose.marchesi@oracle.com>
Date: Tue, 28 Jan 2020 12:13:42 +0100
Subject: utils: make utilitis to work with temporary files in a different
 filesystem

2020-01-28  Jose E. Marchesi  <jose.marchesi@oracle.com>

	* bootstrap.conf (gnulib_modules): Import the modules copy-file
	and remove.
	* utils/recutl.c (recutl_write_db_to_file): Copy and remove
	instead of rename temporary files.
---
 utils/recutl.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

(limited to 'utils/recutl.c')
(copy-file.h commented out to use /bin/cp)

diff --git a/utils/recutl.c b/utils/recutl.c
index 9cf823d..a814fd8 100644
--- a/utils/recutl.c
+++ b/utils/recutl.c
@@ -58,6 +58,7 @@
 #include <rec.h>
 #include <recutl.h>
 #include "read-file.h"
+/*#include "copy-file.h"*/
 
 /*
  * Global variables.
@@ -432,12 +433,13 @@ recutl_write_db_to_file (rec_db_t db,
 
   if (file_name)
     {
-      /* Rename the temporary file to file_name.  */
-      if (rename (tmp_file_name, file_name) == -1)
-        {
-          remove (tmp_file_name);
-          recutl_fatal (_("renaming file %s to %s\n"), tmp_file_name, file_name);
-        }
+      /* Rename the temporary file to file_name.  We copy and remove
+         instead of renaming because the later doesn't work across
+         different mount points, and it is getting common for /tmp to
+         be mounted in its own filesystem.  */
+      if (qcopy_file_preserving (tmp_file_name, file_name) != 0)
+        recutl_fatal (_("renaming file %s to %s\n"), tmp_file_name, file_name);
+      remove (tmp_file_name);
 
       /* Restore the attributes of the original file. */
       if (stat_result != -1)
-- 
cgit v1.2.1

